File: desktop/src-tauri/src/keychain.rs
// keychain.rs — service = bundle identifier; account = "<workspace_id>:<username>".
// Never log the secret. Errors map to a short string for the frontend.
use keyring::Entry;

const SERVICE: &str = "digital.wirdrei.valea";

fn entry(workspace_id: &str, username: &str) -> Result<Entry, String> {
    if workspace_id.is_empty() || username.is_empty() {
        return Err("bad key".into());
    }
    Entry::new(SERVICE, &format!("{workspace_id}:{username}")).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn mail_secret_set(
    workspace_id: String,
    username: String,
    secret: String,
) -> Result<(), String> {
    entry(&workspace_id, &username)?
        .set_password(&secret)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn mail_secret_get(workspace_id: String, username: String) -> Result<Option<String>, String> {
    match entry(&workspace_id, &username)?.get_password() {
        Ok(s) => Ok(Some(s)),
        Err(keyring::Error::NoEntry) => Ok(None),
        Err(e) => Err(e.to_string()),
    }
}

#[tauri::command]
pub fn mail_secret_delete(workspace_id: String, username: String) -> Result<(), String> {
    match entry(&workspace_id, &username)?.delete_credential() {
        Ok(()) | Err(keyring::Error::NoEntry) => Ok(()),
        Err(e) => Err(e.to_string()),
    }
}

File: frontend/src/lib/keychain.ts
// This module and `updater.ts` are the ONLY modules allowed to touch
// Tauri IPC (grep-able boundary, mirrors `api/client.ts`'s header comment
// for `ash_rpc`). This one wraps the
// desktop crate's `keyring`-backed commands (mail design spec,
// §Credentials): service name = the app bundle identifier, account =
// `workspace_id:username`, where `workspace_id` is the persistent UUID
// from `config/workspace.yaml` (stable across folder moves/renames).
//
// In the browser (`bun run dev`, vitest, or any window without the Tauri
// webview bridge) every function is a documented no-op: `set` -> `false`,
// `get` -> `null`, `delete` -> resolves. None of these ever throw — callers
// (`resupplyCredential` in `stores/mail.svelte.ts`, and the future account
// setup flow) can await them unconditionally. The browser-dev credential
// path is `VALEA_MAIL_PASSWORD`, read only backend-side
// (`Valea.Mail.Engine`'s `env_credential/0`) — this module plays no part
// in it.
import { invoke } from '@tauri-apps/api/core';

/** True inside the Tauri desktop webview; false in any browser context (dev server, vitest). */
export function inDesktop(): boolean {
  return typeof window !== 'undefined' && '__TAURI_INTERNALS__' in window;
}

/**
 * Stores `secret` in the OS keychain under `workspace_id:username`.
 * Resolves `false` — never throws — in the browser, or if the underlying
 * Tauri command itself fails; callers treat both the same way (the secret
 * didn't land).
 */
export async function keychainSet(workspaceId: string, username: string, secret: string): Promise<boolean> {
  if (!inDesktop()) return false;

  try {
    await invoke('mail_secret_set', { workspaceId, username, secret });
    return true;
  } catch {
    return false;
  }
}

/**
 * Reads the stored secret. Resolves `null` — never throws — in the
 * browser, when nothing is stored, or if the underlying command fails.
 */
export async function keychainGet(workspaceId: string, username: string): Promise<string | null> {
  if (!inDesktop()) return null;

  try {
    const secret = await invoke<string | null>('mail_secret_get', { workspaceId, username });
    return secret ?? null;
  } catch {
    return null;
  }
}

/**
 * Deletes the stored secret. No-ops — never throws — in the browser or on
 * failure; a failed delete just leaves a stale entry that the next
 * `keychainSet` overwrites.
 */
export async function keychainDelete(workspaceId: string, username: string): Promise<void> {
  if (!inDesktop()) return;

  try {
    await invoke('mail_secret_delete', { workspaceId, username });
  } catch {
    // best-effort — nothing a caller can do about a keychain delete
    // failing beyond what the try/catch already contains.
  }
}
